### ISSUE TITLE
Extracting inherited annotations may face compiler bug

### ISSUE THREAD

**Issue by danielleontiev** [propensive/magnolia#385]
Me and @vladislavsheludchenkov were updating tapir on `0.20.0-M9` version (the one that brings magnolia `1.1.0`) and have faced strange behavior. `Schema` derivation have broken on some classes in our project with intriguing error:

```
[error] Main.scala:8:17: class type required but <notype> found
[error]   Schema.derived[Foo]
```

Reproducer for that can be found [here](https://github.com/danielleontiev/magnolia-annotations-bug), run `sbt ";clean;tapir/compile"` in terminal.

Untangling the reason for the error we have finally found the unexpected compiler behavior: https://github.com/scala/bug/issues/12536

The behavior is the following - sometimes (probably, it depends on compilation order), weird annotations are visible to macro that can fail compilation if included in the result tree (which is the case for magnolia).

Before https://github.com/softwaremill/magnolia/pull/376, magnolia got all annotation from constructor members (which is correct, because annotations on class fields should end up on constructor parameters, if they are declared in the constructor [proof](https://www.scala-lang.org/api/2.13.8/scala/annotation/meta/index.html)). Now, taking into account the changes in PR, magnolia starts to get annotations from methods and fields too. When method `annotationsOf` in `magnolia.scala` is called on constructor parameter, under the hood it climbs to the class and calls `.members` on it - result includes methods and fields. Together with the compiler which puts weird annotations on fields and methods sometimes, it definitely leads to hours of debugging 😃.

For example, if we have `case class Foo(@description("foo!") foo: String)` and we are running under the conditions that trigger compiler to put annotations on fields and methods, it will put them on `foo` field and `foo` getter. So, calling `annotationsOf(<constructor parameter foo>)` will result in the following under the hood: `.owner` called returning constructor -> `.owner` called again returning class `Foo` -> `.baseClasses` called returning `java.lang.Object`, some traits from `scala.` and, which is important, the `Foo` class itself - together with unwanted annotations. Next, these annotations are included in the resulting list that fails compilation on the later steps.

It's tempting to create a workaround by simply filtering out the class itself from the output of `.baseClasses`, but if we have the following setup:

```scala
class Base(
    @description("base foo")
    val foo: String
)
case class Foo(
    @description("foo!")
    override val foo: String
) extends Base(foo)
```

compiler puts unwanted annotations inside both `Foo` and `Base`, so, filtering out `Foo` won't help. Runnable example [here](https://github.com/danielleontiev/magnolia-annotations-bug/blob/main/generic/src/main/scala/Gen.scala#L21).

So, the fix may include typechecking the tree and discarding an annotation if it doesn't work, for example.

Meanwhile, some issues in the implementation of `annotationsOf` have been found.

1. Pattern matching does not work as expected because `type MethodSymbol >: Null <: MethodSymbolApi with TermSymbol`

Runnable [example](https://github.com/danielleontiev/magnolia-annotations-bug/blob/main/pattern-matching/src/main/scala/Main.scala)

2. Comparing names of symbols looks really unreliable

When annotation is added to class field (and not the constructor), it is renamed by compiler and comparing names won't work. [example](https://github.com/danielleontiev/magnolia-annotations-bug/blob/main/names/src/main/scala/Main.scala#L12-L15)

3. The algorithm does not check parent's constructor parameters

`annotationsOf` only checks methods and fields of the class and it doesn't take parent class constructor parameters into account, where parent annotation could live if they declared on constructor parameters.

Thank you @vladislavsheludchenkov for spending time chasing it!

**Comment by adamw** [comment]
cc @kubinio123 

My first idea would be to read only constructor annotations, but maybe that's too simplistic and causes other problems? :)

**Comment by danielleontiev** [comment]
So, it depends on the goals of extracting inherited annotations. For example, reading only constructor annotations will catch annotation in this case:

```scala
class Base(
    @description("base foo")
    val foo: String
)
case class Foo(
    @description("foo!")
    override val foo: String
) extends Base(foo)
```

but will skip it if base class is declared like this:

```scala
class Base {
    @description("foo!") val foo = 321
  }
```

So, if the second case is rare or not important then your fix can be used, but does the second case really not important?

Moreover, it won't work with `trait`s as they do not have constructors

**Comment by adamw** [comment]
Right, reading annotations from base classes/traits is the main purpose of this change. So I guess trying to run a typecheck might be our best option

**Comment by kubinio123** [comment]
@danielleontiev @OrangeMemes thanks for the great reproducer!

So instead of running a typecheck I went for filtering out `NoType` annotations and linking to a compiler bug.

As for `annotationsOf` algorithm itself I fixed differentiating methods and terms and added inheriting annotations from base classes constructor parameters.

I'am not a fan of comparing symbols by `name` but I don't know a better way to do it. Also in your reproducer you are looking at the `encodedName` of members while the actual `name` remains the same for both.
